engine/src/main/java/io/zeebe/engine/nwe/container/MultiInstanceBodyProcessor.java

```diff
@@ -140,15 +140,86 @@ public void onCompleted(
 
   @Override
   public void onTerminating(
-      final ExecutableMultiInstanceBody element, final BpmnElementContext context) {}
+      final ExecutableMultiInstanceBody element, final BpmnElementContext context) {
+
+    eventSubscriptionBehavior.unsubscribeFromEvents(context);
+
+    stateTransitionBehavior.terminateChildInstances(context);
+  }
 
   @Override
   public void onTerminated(
-      final ExecutableMultiInstanceBody element, final BpmnElementContext context) {}
+      final ExecutableMultiInstanceBody element, final BpmnElementContext context) {
+
+    eventSubscriptionBehavior.publishTriggeredBoundaryEvent(context);
+
+    incidentBehavior.resolveIncidents(context);
+
+    stateTransitionBehavior.onElementTerminated(element, context);
+
+    stateBehavior.consumeToken(context);
+  }
 
   @Override
   public void onEventOccurred(
-      final ExecutableMultiInstanceBody element, final BpmnElementContext context) {}
+      final ExecutableMultiInstanceBody element, final BpmnElementContext context) {
+
+    eventSubscriptionBehavior.triggerBoundaryEvent(element, context);
+  }
+
+  @Override
+  public void onChildCompleted(
+      final ExecutableMultiInstanceBody element,
+      final BpmnElementContext flowScopeContext,
+      final BpmnElementContext childContext) {
+    final var loopCharacteristics = element.getLoopCharacteristics();
+
+    if (loopCharacteristics.isSequential()) {
+
+      final var inputCollectionVariable = readInputCollectionVariable(element, childContext);
+      if (inputCollectionVariable.isEmpty()) {
+        return;
+      }
+
+      final var array = inputCollectionVariable.get();
+      final var loopCounter =
+          stateBehavior.getFlowScopeInstance(childContext).getMultiInstanceLoopCounter();
+      if (loopCounter < array.size()) {
+
+        final var item = array.get(loopCounter);
+        createInnerInstance(element, flowScopeContext, item);
+      }
+    }
+
+    final Optional<Boolean> updatedSuccessfully =
+        loopCharacteristics
+            .getOutputCollection()
+            .map(variableName -> updateOutputCollection(element, childContext, variableName));
+
+    if (updatedSuccessfully.isPresent() && !updatedSuccessfully.get()) {
+      // An incident was raised while updating the output collection, stop handling activity
```

Comment:
Perhaps this is also the time to refactor to `Either<Failure,OutputCollection>` or something similar.

Reply:
Good point. I ignored this part :sweat_smile: